pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/SwitchDensityRule.java

```diff
@@ -36,51 +37,32 @@
                        .build();
 
     public SwitchDensityRule() {
-        super(ASTSwitchStatement.class);
+        super(ASTSwitchStatement.class, ASTSwitchExpression.class);
         definePropertyDescriptor(REPORT_LEVEL);
     }
 
     @Override
     public Object visit(ASTSwitchStatement node, Object data) {
-        double density = new SwitchDensityVisitor().compute(node);
-        if (density >= getProperty(REPORT_LEVEL)) {
-            addViolation(data, node);
-        }
-        return super.visit(node, data);
+        return visitSwitchLike(node, data);
     }
 
-    private static class SwitchDensityVisitor extends JavaParserVisitorAdapter {
-
-        private int labels = 0;
-        private int stmts = 0;
-        private ASTSwitchStatement root;
-
-
-        double compute(ASTSwitchStatement root) {
-            this.root = root;
-            root.jjtAccept(this, null);
-            return labels == 0 ? 0 : ((double) stmts) / labels;
-        }
-
+    @Override
+    public Object visit(ASTSwitchExpression node, Object data) {
+        return visitSwitchLike(node, data);
+    }
 
-        @Override
-        public Object visitStatement(ASTStatement statement, Object data) {
-            stmts++;
-            return super.visitStatement(statement, data);
-        }
+    public Void visitSwitchLike(ASTSwitchLike node, Object data) {
+        // note: this does not cross find boundaries.
+        int stmtCount = node.descendants(ASTStatement.class).count();
+        int labelCount = node.getBranches()
+                .map(ASTSwitchBranch::getLabel)
+                .sumBy(label -> label.isDefault() ? 1 : label.getExprList().count());
 
-        @Override
-        public Object visit(ASTExpression node, Object data) {
-            // don't recurse on anonymous class, etc
-            return data;
-        }
-
-        @Override
-        public Object visit(ASTSwitchLabel switchLabel, Object data) {
-            if (switchLabel.getParent() == root) {
-                labels++;
-            }
-            return super.visit(switchLabel, data);
+        // note: if labelCount is zero, double division will produce NaN, not ArithmeticException
```

Comment:
```suggestion
        // note: if labelCount is zero, double division will produce +Infinity or NaN, not ArithmeticException
```
+Infinity if stmtCount > 0 and labelCount is zero, NaN if both stmtCount and labelCount is zero